### PR TITLE
fu-progressbar: Show a time estimate after update has started

### PR DESCRIPTION
**Pull request by superm1**
For many devices this won't matter, but for some devices I've experienced flash times that are nominally 5 minutes.  I also have one device that takes over 40 minutes to flash (although I hope that we can optimize that better...)

### CODE REVIEW

**Comment by hughsie** [comment]
In concept, yes. I do think we need to only show the time estimate after a few seconds have elapsed, and when we have some degree of confidence in the prediction. We'll get laughed out of the room if we say things like "2 hours remaining" -> "2 minutes remaining" -> "15 minutes remaining".

For a 15 minute flash (which is a huge amount of time btw) we'd only get a percentage changed event every 9 seconds, so the logic of showing the remaining time after say 20 seconds elapsed is probably a good metric. I think we'd just need to ensure that the time difference per-update is proportional (i.e. 0,1,2%), even if we have a 50% difference between updates.

**Comment by superm1** [comment]
In terms of waiting until we have confidence in the display, one thought is to calculate on every call to the function and only start display when the remaining time estimate begins to decrease.

File: src/fu-progressbar.c [hunk 1]
```diff
@@ -24,6 +25,7 @@ struct _FuProgressbar
 	guint			 to_erase;		/* chars */
 	guint			 timer_id;
 	gint64			 last_animated;		/* monotonic */
+	GTimer			 *time_elapsed;
```
Comment (hughsie) [inline src/fu-progressbar.c]: No space before the `*` pls

File: src/fu-progressbar.c [hunk 2]
```diff
@@ -103,6 +105,7 @@ fu_progressbar_refresh (FuProgressbar *self, FwupdStatus status, guint percentag
 	guint i;
 	gboolean is_idle_newline = FALSE;
 	g_autoptr(GString) str = g_string_new (NULL);
+	g_autofree gchar *time_remaining = NULL;
```
Comment (hughsie) [inline src/fu-progressbar.c]: This can move inside the `if` block, no?

File: src/fu-progressbar.c [hunk 3]
```diff
@@ -135,6 +138,16 @@ fu_progressbar_refresh (FuProgressbar *self, FwupdStatus status, guint percentag
 	}
 	g_string_append_c (str, ']');
 
+	/* once we have some data (10%) show an estimate of time remaining */
+	if (percentage < 100 && percentage > 10) {
+		gdouble rate = percentage / g_timer_elapsed (self->time_elapsed, NULL);
+		time_remaining = g_strdup_printf ("%.2f", (100-percentage) * rate);
```
Comment (hughsie) [inline src/fu-progressbar.c]: spaces around the `-` pls, and I think you might want to cast the result to float, e.g. using `100.f`

File: src/fu-progressbar.c [hunk 4]
```diff
@@ -135,6 +138,16 @@ fu_progressbar_refresh (FuProgressbar *self, FwupdStatus status, guint percentag
...
+		g_string_append_c (str, ' ');
+		g_string_append (str, _("Estimated time remaining:"));
```
Comment (hughsie) [inline src/fu-progressbar.c]: Needs a translator comment explaining the context

File: src/fu-progressbar.c (hunk 1 above)
Comment (superm1) [reply src/fu-progressbar.c]: fixed

File: src/fu-progressbar.c (hunk 2 above)
Comment (superm1) [reply src/fu-progressbar.c]: fixed

File: src/fu-progressbar.c (hunk 3 above)
Comment (superm1) [reply src/fu-progressbar.c]: fixed (although not with casting)

File: src/fu-progressbar.c (hunk 4 above)
Comment (superm1) [reply src/fu-progressbar.c]: added

**Comment by superm1** [comment]
Rather than a random number (10 or 20 seconds elapsed) I think comparing the estimates works better, once they start to decrease we should have good confidence it's hit steady state.

**Comment by superm1** [comment]
Any more thoughts on this PR now?

**Comment by hughsie** [comment]
Yes, sorry, LGTM.

**Comment by superm1** [comment]
Just realized I didn't hit the squash button; sorry.  So we've got a commit in master `fixups` unless you want me to force push master to fix it.

**Comment by hughsie** [comment]
Don't worry about the commit history; next time. Force pushing master will probably break the world. :)

**Comment by superm1** [comment]
OK thanks.

**Pull request merged**